In [128]:
import warnings
warnings.filterwarnings('ignore')

In [129]:
import numpy as np
import pandas as pd
from math import sqrt

In [130]:
pd.set_option('display.max_rows', 100)
pd.set_option('display.max_columns', 100)
pd.set_option('display.width', -1)

In [131]:
infile = './transformed_data/train_heuristic.csv'

In [132]:
df = pd.read_csv(infile)
df.head()

OverallQual  \
0            7   
1            6   
2            7   
3            7   
4            8   

   TotalArea  \
0       4276   
1       3786   
2       4492   
3       4190   
4       5541   

   GarageCars  \
0           2   
1           2   
2           2   
3           3   
4           3   

   PropertyAge  \
0            5   
1           31   
2            7   
3           91   
4            8   

   MSSubClass  \
0          60   
1          20   
2          60   
3          70   
4          60   

  MSZoning  \
0       RL   
1       RL   
2       RL   
3       RL   
4       RL   

   Neighborhood  \
0             1   
1             2   
2             1   
3             3   
4             4   

   HasPorch  \
0         1   
1         0   
2         1   
3         1   
4         1   

   HasDeck  \
0        0   
1        1   
2        0   
3        0   
4        1   

  LotConfig  \
0    Inside   
1       FR2   
2    Inside   
3    Corner   
4       FR2   

  HouseStyle  \
0     2Story   
1     1Story   
2     2Story   
3     2Story   
4     2Story   

  SaleType  \
0       WD   
1       WD   
2       WD   
3       WD   
4       WD   

  SaleCondition  \
0        Normal   
1        Normal   
2        Normal   
3       Abnorml   
4        Normal   

   SalePrice  
0     208500  
1     181500  
2     223500  
3     140000  
4     250000

In [133]:
# Taking the mean of the best predictor and using that as the predictor

tab = df[['OverallQual', 'SalePrice']].groupby(['OverallQual']).mean().reset_index()
tab.rename(columns={'SalePrice':'SalePricePred'}, inplace=True)

In [134]:
df = pd.merge(
    left=df,
    right=tab,
    left_on='OverallQual',
    right_on='OverallQual',
    how='left'
)

In [135]:
def RMSE(row):
    diff_sq = sqrt((row['SalePrice'] - row['SalePricePred'])**2)
    return diff_sq

df['RMSE'] = df.apply(RMSE, axis=1)
df['RMSE_log'] = np.log(df['RMSE'])
result = df['RMSE_log'].mean()
result

9.800078624737752

9.8 is already enough not to come last! Go heuristics!